cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -41,11 +41,115 @@ static int sim_ending = 0;
 
 #include "simulatormodule.h"
 #include <cocotb_utils.h>     // COCOTB_UNUSED
+#include <type_traits>
+
 
 struct module_state {
     PyObject *error;
 };
 
+
+/* define the extension types as templates */
+namespace {
+    template<typename gpi_hdl>
+    struct gpi_hdl_Object {
+        PyObject_HEAD
+        gpi_hdl hdl;
+
+        // The python type object, in a place that is easy to retrieve in templates
+        static PyTypeObject py_type;
+    };
+
+    template<typename gpi_hdl>
+    static PyObject *gpi_hdl_repr(gpi_hdl_Object<gpi_hdl> *self) {
+        PyTypeObject *type = Py_TYPE(self);
+        return PyUnicode_FromFormat("<%s at %p>",
+                                      type->tp_name, self->hdl);
+    }
+    template<typename gpi_hdl>
+    static Py_hash_t gpi_hdl_hash(gpi_hdl_Object<gpi_hdl> *self) {
+        Py_hash_t ret = (Py_hash_t)self->hdl;
+        // hash must never return -1
+        if (ret == (Py_hash_t)-1) {
+            ret = (Py_hash_t)-2;
+        }
+        return ret;
+    }
+
+    template<typename gpi_hdl>
+    static PyObject *gpi_hdl_New(gpi_hdl hdl) {
+        if (hdl == NULL) {
+            Py_RETURN_NONE;
+        }
+        auto *obj = PyObject_New(gpi_hdl_Object<gpi_hdl>, &gpi_hdl_Object<gpi_hdl>::py_type);
+        if (obj == NULL) {
+            return NULL;
+        }
+        obj->hdl = hdl;
+        return (PyObject *)obj;
+    }
+
+    template<typename gpi_hdl>
+    static PyObject *gpi_hdl_richcompare(PyObject *self, PyObject *other, int op) {
+        if (
+                Py_TYPE(self) != &gpi_hdl_Object<gpi_hdl>::py_type ||
+                Py_TYPE(other) != &gpi_hdl_Object<gpi_hdl>::py_type) {
+            Py_RETURN_NOTIMPLEMENTED;
+        }
+
+        auto self_hdl_obj = (gpi_hdl_Object<gpi_hdl> *)self;
+        auto other_hdl_obj = (gpi_hdl_Object<gpi_hdl> *)other;
+
+        switch (op) {
+            case Py_EQ: return PyBool_FromLong(self_hdl_obj->hdl == other_hdl_obj->hdl);
+            case Py_NE: return PyBool_FromLong(self_hdl_obj->hdl != other_hdl_obj->hdl);
+            default: Py_RETURN_NOTIMPLEMENTED;
+        }
+    }
+
+    // Initialize the python type slots
+    template<typename gpi_hdl>
+    PyTypeObject fill_common_slots() {
+        PyTypeObject type;
+        type.ob_base = {PyObject_HEAD_INIT(NULL) 0};
+        type.tp_basicsize = sizeof(gpi_hdl_Object<gpi_hdl>);
+        type.tp_repr = (reprfunc)gpi_hdl_repr<gpi_hdl>;
+        type.tp_hash = (hashfunc)gpi_hdl_hash<gpi_hdl>;
```

Comment:
It would be nice if you didn't have to cast this. Kind of makes using `gpi_hdl_Object` in the function pointless.

Reply:
I don't agree, it makes the function body simpler. At any rate, it's convention to do so in python.

Reply:
The C++ standard (and probably the C standard as well), state that it's UB. Not that I expect it to do the wrong thing, but I tend to shy away from it if possible.

Reply:
In that case, the python runtime is already full of the same UB. I think matching the python docs is more valuable than avoiding UB we already rely on elsewhere.